lib/pbench/agent/base.py

```diff
@@ -21,7 +21,10 @@ def __init__(self, context):
         self.config = PbenchAgentConfig(self.context.config)
         self.name = os.path.basename(sys.argv[0])
 
-        self.pbench_run = self.config.pbench_run
+        env_pbench_run = os.environ.get("pbench_run")
+        self.pbench_run = pathlib.Path(
+            env_pbench_run if env_pbench_run else self.config.pbench_run
+        )
```

Comment:
This mostly matches the code in `agent/base`, which is used by `pbench-register-tool`. The only real difference is that `base`, interestingly, verifies existence of the directory only if it was provided by a `pbench_run` environment variable.

My concern is exactly where we expect the default `/var/lib/pbench-agent/` to be created. The Python code expects it to already exist, but the following bash `base` statement is a `mkdir -p "${pbench_run}/tmp" which would create it if it doesn't already exist ... the Python code will instead fail on the following `if not self.pbench_run.exists()`.

Then again, since this is mostly pre-existing code I assume we'd have run into this problem already if there were realistic cases where `/var/lib/pbench-agent` *didn't* exist, and it probably doesn't matter.

Adding a click CLI unit test case for your new code, however, would be a good idea.

Reply:
So, what should we do? Should I create the directory if it is not present? or let it throw error? 

if [[ -z "$pbench_run" ]]; then                                                                                                       1   ┊ pbench_run=$(pbench-config pbench_run pbench-agent)
  2   ┊ if [[ -z "$pbench_run" ]]; then
  3   ┊ ┊ ┊ pbench_run=/var/lib/pbench-agent
  4   ┊ fi
  5   ┊ export pbench_run
  6 else
  7   ┊ if [[ ! -d $pbench_run ]]; then
  8   ┊ ┊ ┊ echo "[ERROR] the provided pbench run directory, ${pbench_run}, does not exist" >&2
  9   ┊ ┊ ┊ exit 1
 10   ┊ fi
 11 fi


In `agent/base` it does check if the directory exists and if it doesn't then it exits and doesn't continue.

Reply:
Yes; the difference is that the bash version only exits with error if `pbench_run` is defined to a non-blank value and the directory doesn't exist, whereas the Python makes the check even if `pbench_run` wasn't defined.

Whether this matters depends on whether there's supposed to be a guarantee that the `pbench-agent.cfg` directory already exists; and whether the bash `mkdir -p` that immediately follows is a deliberate attempt to dynamically create a "default" `pbench_run` directory or an accident. (That is, `mkdir ${pbench_run}/tmp` would have worked just fine if we were sure that `${pbench_run}` already existed, but the `-p` will create the entire path if necessary.)

Also, if this is a bug, it wasn't added by your changes, so the issue is somewhat separate from your PR even though it's adjacent code. (You added the environment variable lookup, but the config file default path isn't affected.) Maybe @portante or @ndokos can clarify the historical intent without requiring someone to search through the code to figure out whether a missing default `pbench_run` directory at this point ought to be an error.

Reply:
First, I believe the Bash code is wrong, it should not issue an `mkdir -p ${pbench_run}/tmp`.

The check for existence of `${pbench_run}` should be removed entirely, and a `mkdir ${pbench_run}/tmp` should be used, where on failure we report why it failed.

That is probably best addressed in a separate PR.

Once that is addressed, then I believe we can resolve cleanly the differences between the Python3 code and the Bash code.

Reply:
> That is probably best addressed in a separate PR.

By which, @portante, means please go do that one first, and then come back and build this PR on that one.

Reply:
Under the circumstances, we definitely need an issue/Jira to track the correct behavior and get the bash and Python versions completely aligned... however I think it's also perfectly reasonable to add the `_pbench_run` check here and *then* fix the check/create paths in both afterwards. (This change doesn't make the disconnect worse.)

Reply:
> we definitely need an issue/Jira to track the correct behavior and get the bash and Python versions completely aligned...

Not to go out on a tangent or anything, but I believe the plan is to remove the Bash code and replace it with Python, at which point they presumably will be completely aligned (and there _is_ a Jira ticket (an epic, I think) for that...).  🙂

Reply:
Tangent? Us? When have we *ever* gone on a tangent about any discussion. Name one example. 👀 OK, name 1,000 examples... 

But yes, that's certainly true; we also explicitly kicked the remaining "Pythonization" out of sprint 8 and Agent 0.71 so we want to do what we can prior to that.

I fussed just a bit yesterday afternoon over how to organize something like an epic, perhaps, for "0.71 completers" which might or might not include defects... but didn't really strike any inspiration.